### PR TITLE
Gesture manager (initial)

### PR DESCRIPTION
**Pull request by robert00s**
This PR creates a gesture manager for now for 3 gestures and a limited number of actions.
I will try to expand the manager in the future.
Manager supports separate gestures for the file manager and the reader. Gestures from the manager have a higher priority than built-in gestures.

Settings available in Gear -> Device -> Gesture manager

a - koreader_173
a - koreader_174
a - koreader_175
kontratyp - koreader_177

Close: #1008

### CODE REVIEW

**Comment by KenMaltby** [comment]
Minor question, how is "Disable" being used? Is it an option to disable that touch area?  What is being disabled in each menu?

**Comment by robert00s** [comment]
@KenMaltby 
Disable is default for all touch area (Gestures as they are now and nothing changes). 
If you want to switch to night mode after touching right bottom corner of the screen, you choose the `Night mode` in submenu `Tap right bottom corner`.  If you want to cancel this action, you choose `Disable`.

**Comment by robert00s** [comment]
This gestures work only in FileManager and Reader mode. They don't work when another widget is opened (e.g. table of content, history - widgets have own set of predefined gesture).

**Comment by Frenzie** [comment]
Most interesting. :-)

**Comment by KenMaltby** [comment]
It looks like there would be room for the check box line to be "No Manager Option", checked by default.  Perhaps "Radio Buttons" would work better?

**Comment by AlanSP1** [comment]
Is there somewhere info on all gestures? I think those in wiki are not up to date? Or I'm mistaken?

**Comment by Frenzie** [comment]
There's the swipe to open/close menu and @poire-z just added swipe down to close some popups.

Is there anything else that's been added?

**Comment by poire-z** [comment]
> Is there anything else that's been added?

There is also pinch/spread in reader to change font size, #2768.

> Is there somewhere info on all gestures?

I guess we're just mentionning the gestures that work in reader/filemanager, because there are many other that are widget specifics (Dictionary, TextViewer, TextEditor, ImageViewer, MovableContainer have their own set of gestures, that I think are already mentionned in their proper wiki page).

**Comment by poire-z** [comment]
> It looks like there would be room for the check box line to be "No Manager Option", checked by default. Perhaps "Radio Buttons" would work better?

May be each action menu could just be organised differently, and avoid the word `Disable`:

```
Default action
--------------
No action
--------------
All the other new actions
```
eg, for Short diagonal swipe:
```
Full screen refresh (default)
--------------
Nothing
--------------
Folder up
Toggle nightmode
Reading progress
```

File: frontend/ui/widget/menu.lua [hunk 1]
```diff
@@ -698,6 +698,7 @@ function Menu:init()
         callback = function() self:onReturn() end,
         bordersize = 0,
         show_parent = self,
+        readonly = self.return_arrow_propagation,
```
Comment (poire-z) [inline frontend/ui/widget/menu.lua]: So, this return button will not work anymore? I don't see it being re-enabled.
I see it being used in the Statistics navigation. Anywhere else?

File: frontend/apps/reader/modules/readergesture.lua [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        end,
+        callback = function()
+            gesture_manager[ges] = action
+            G_reader_settings:saveSetting(self.ges_mode, gesture_manager)
+            self:subMenuAction(ges, action)
+        end,
+    }
+end
+
+function ReaderGesture:subMenuAction(ges, action)
```
Comment (poire-z) [inline frontend/apps/reader/modules/readergesture.lua]: This function name seems wrong, and should probably be renamed to `ReaderGesture:setupGesture(ges, action)`

**Comment by poire-z** [comment]
Looks quite unintrusive to the other code, which is nice :)
(Can't comment about the touchzone/dependancy graph code, as I never really looked at it, so I trust you on this.)

File: frontend/ui/widget/menu.lua (hunk 1 above)
Comment (robert00s) [reply frontend/ui/widget/menu.lua]: Yes, It's also used in OPDS browser.

**Comment by AlanSP1** [comment]
OK, I updated wiki gestures page with gestures you mentioned and I know of. If there's some more gestures, please mention them.

**Comment by robert00s** [comment]
* new menu:
pragnienie - koreader_035
* refactoring code
* add default gesture
* add toggle frontlight gesture
* toggle frontlight gesture move from kobolight plugin to gesture manager
* prevent crash KO after choose reading progress and disable plugin

File: frontend/ui/widget/menu.lua [hunk 3]
```diff
@@ -1187,7 +1188,9 @@ function Menu:onSwipe(arg, ges_ev)
     elseif ges_ev.direction == "north" then
         -- no use for now
         do end -- luacheck: ignore 541
-    else -- diagonal swipe
+    elseif not G_reader_settings:readSetting("gesture_fm")["short_diagonal_swipe"] then
```
[the remark this answers is not in the document]
Comment (robert00s) [reply frontend/ui/widget/menu.lua]: Good idea. Thanks :)

File: frontend/apps/reader/modules/readergesture.lua [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    elseif action =="folder_up" then
+        self.ui.file_chooser:changeToPath(string.format("%s/..", self.ui.file_chooser.path))
+    elseif action =="toggle_frontlight" then
+        Device:getPowerDevice():toggleFrontlight()
+        self:onShowFLOnOff()
+    end
+    if return_value then
+        -- a long diagonal swipe may also be used for taking a screenshot,
+        -- so let it propagate
+        -- no implemented for now
```
[the remark this answers is not in the document]
Comment (robert00s) [reply frontend/apps/reader/modules/readergesture.lua]: I had another conception a few days ago. This is no more needed so I will remove return_value :)

File: frontend/apps/reader/modules/readergesture.lua [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+function ReaderGesture:buildMenu(ges, default)
+    local gesture_manager = G_reader_settings:readSetting(self.ges_mode)
+    local menu = {
+        --{_("Menu element), "action", enable_element},
+        {_("Nothing"), "nothing", true },
+        {_("Backward 10 pages"), "page_update_down10", not self.is_docless},
+        {_("Bookmarks"), "bookmarks", not self.is_docless},
+        {_("Folder up"), "folder_up", self.is_docless},
+        {_("Forward 10 pages"), "page_update_up10", not self.is_docless},
+        {_("Full screen refresh"), "full_refresh", true},
```
[the remark this answers is not in the document]
Comment (robert00s) [reply frontend/apps/reader/modules/readergesture.lua]: Agree. Logical sorting is better :)

File: plugins/kobolight.koplugin/main.lua [hunk 6]
```diff
@@ -165,12 +154,6 @@ function KoboLight:onShowOnOff()
     return true
 end
 
-function KoboLight:onTap()
-    Device:getPowerDevice():toggleFrontlight()
-    self:onShowOnOff()
-    return true
-end
-
```
[the remark this answers is not in the document]
Comment (robert00s) [reply plugins/kobolight.koplugin/main.lua]: I want to migrate all kobolight (Frontlight gesture controller) to readerfrontlight + readergesture and remove completely this plugin. This is first step (tap), next will be swipe left and right side of the screen and all logic will be moved to readerfrontlight.
I want to do that on another PR when new gestures will be added.

**Comment by AlanSP1** [comment]
Not sure if small typo is corrected, it is "table of content" not "context".

**Comment by robert00s** [comment]
@AlanSP1 
It was corrected yet :)

File: frontend/apps/reader/modules/readergesture.lua [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+        },
+    }
+end
+
+function ReaderGesture:buildMenu(ges, default)
+    local gesture_manager = G_reader_settings:readSetting(self.ges_mode)
+    local menu = {
+        --{_("Menu element), "action", enable_element},
+        {_("Nothing"), "nothing", true },
+        {_("Backward 10 pages"), "page_update_down10", not self.is_docless},
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readergesture.lua]: Just `Back 10 pages`.

**Comment by Frenzie** [comment]
@robert00s Please merge when ready, unless @poire-z has any further comments. :-)

**Comment by Frenzie** [comment]
Ah, I didn't read it in detail because you already did.

**Comment by robert00s** [comment]
All work is done. Long diagonal swipe still take screenshots.

**Comment by poire-z** [comment]
2 small issues:
Now that ReaderGesture handles diagonal swipe, the original code is not used, so we may miss the additional stuff it does. For the reader, it additionally updated the footer:
https://github.com/koreader/koreader/blob/1e275fdc4f2e862115191eec2146025a7ed964c8/frontend/apps/reader/modules/readerrolling.lua#L376-L381
We can quickly get it back by adding to readerui.lua:
```lua
    if Device:isTouchDevice() then
        -- gesture manager
        self:registerModule("gesture", ReaderGesture:new {
            document = self.document,
+           view = self.view,
            ui = self,
        })
    end
```
and to readergesture.lua:
```lua
    elseif action == "full_refresh" then
+       if self.view then
+           -- update footer (time & battery)
+           self.view.footer:updateFooter()
+       end
        UIManager:setDirty(nil, "full")
```
But I don't know, may be it would be more correct to just not handle the action when it's the default action, and to just let the event propagate so it's handled by the original code (and we're not confused when we add some stuff to readerrolling and see it not happening because it's actually readergesture that is now handling the gesture).

Also (not investigated, was just testing), when setting bottom left to toggle night mode, night mode is not immediately set: you need to change page or trigger full refresh to see it.

**Comment by robert00s** [comment]
@poire-z Thanks.

> But I don't know, may be it would be more correct to just not handle the action when it's the default action, and to just let the event propagate so it's handled by the original code (and we're not confused when we add some stuff to readerrolling and see it not happening because it's actually readergesture that is now handling the gesture).


I think that your code is better than handle refresh by the original code. This would be (maybe) better for digonal swipe but not for tap (left or right corner). Without your patch when we select full refresh by tapping in corner we don't have update footer.

> Also (not investigated, was just testing), when setting bottom left to toggle night mode, night mode is not immediately set: you need to change page or trigger full refresh to see it.

I check that later and prepare patch if needed.

**Comment by poire-z** [comment]
Seems like using a hack like this in the above code solves the issue:
```lua
    else -- diagonal swipe
        if self.show_parent and self.show_parent.tapPlus and -- it is FileManager
            G_reader_settings:readSetting("gesture_fm") and G_reader_settings:readSetting("gesture_fm")["short_diagonal_swipe"] then
            -- managed by gesture manager
            do end -- luacheck: ignore 541
        else
            -- trigger full refresh
            UIManager:setDirty(nil, "full")
        end
    end
```
We should probably make that more explicit by just adding an attribute in the FileChooser created by FileManager: `i_am_filemanager = true` :)

**Comment by poire-z** [comment]
(diagonal refresh fix added in #4378)

**Pull request merged**